homeassistant/components/schedule/__init__.py

```diff
@@ -344,6 +352,14 @@ def _update(self, _: datetime | None = None) -> None:
         self._attr_extra_state_attributes = {
             ATTR_NEXT_EVENT: next_event,
         }
+
+        if isinstance(current_data, dict):
+            # Add each key/value pair in the data to the entity's state attributes
+            self._attr_extra_state_attributes.update(current_data)
```

Comment:
This one is a little confusing IMHO, sometimes it is added to the attributes, or otherwise it is stored in a `data` key.

IMHO, this is magic we should avoid. Let's either always store it in a data key (or not).

Honestly, I think we should always use the data key, mostly because it would allow us to exclude it from the recorder (which is an change that should be added to this PR as well).

Reply:
I like the idea of excluding a single key from the recorder history, but my goal with the "flattening" of data attributes into the entity's state attributes was to make it easier to build automations. How do you select, for example, `data.brightness` in an automation trigger?

![entity attributes](https://github.com/home-assistant/core/assets/1693960/a2b9cd6c-0bd4-4872-9c14-06ab6ca3b498)

![automation trigger](https://github.com/home-assistant/core/assets/1693960/515188cc-368d-4807-b59a-77bd84183008)

Reply:
Accessing those is harder indeed. Alternatively, we do flatten them, but exclude all attributes from being recorded in that case.

Reply:
I just pushed a new commit that adds any custom attributes present to `_unrecorded_attributes`. Is that the preferred way to do this?

Reply:
That won't work, you'll need to also set the `Entity` class' private attribute `__combined_unrecorded_attributes`:
```
    self._unrecorded_attributes = frozenset(self.all_custom_data_keys())
    self._Entity__combined_unrecorded_attributes = (
            self._entity_component_unrecorded_attributes | self._unrecorded_attributes
        )
```

Please add a test which ensures the exclusion actually works, search for `test_exclude_attributes` and you'll find tests which do just that.

Reply:
The test update is not correct the state set is just:
```
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=schedule.test, old_state=None, new_state=<state schedule.test=off; editable=False, next_event=2024-08-03T09:00:00-07:00, icon=mdi:party-popper, friendly_name=Party mode @ 2024-08-02T08:51:58.895820-07:00>>
```

The test should assert the state has `CONF_DATA` and `"party_level"` state attributes, and then check those state attributes are not recorded.

Reply:
I think this approach is valid (or at least, it also appears at https://github.com/home-assistant/core/blob/dev/tests/components/select/test_recorder.py#L49).

I just pushed a new commit that updates the Entity class's unrecorded attributes like you suggested and adds assertions that the attributes are present on the entity to the test (in addition to the existing assertions that it is not present in the recorded states).

Reply:
I've tried a few different ways of mocking the current time that I found in other tests and none of them seem to work.

I've tried this:
```python
async_fire_time_changed(hass, datetime(2024, 8, 2, 7, 30))
await hass.async_block_till_done()
```

And this:
```python
with freeze_time(datetime(2024, 8, 2, 7, 30)):
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
```

(both with and without `tzinfo=dt_util.UTC`)

But following either of those methods with this:

```python
state = hass.states.get("schedule.test")
print(state)
```

...results in this output, which says Party mode @ the wrong time:

```
<state schedule.test=off; editable=False, next_event=2024-08-07T03:00:00-07:00, icon=mdi:party-popper, friendly_name=Party mode @ 2024-08-06T23:22:40.040380-07:00>
```

Is there any documentation on mocking times like this? I couldn't find any, and there seems to be no consensus among the other time-based tests that I can make sense of. Thank you for any help you can offer :)

Reply:
Passing a datetime object to `async_fire_time_changed` is the old way to implement time travelling in tests, it is very hard to get right because `async_fire_time_changed` only patches a few select time related functions.

Using freezegun is much preferred because it is much less likely to work since freezegun searches all modules for imported time functions and patched them.

When you use freezegun, you need to distinguish between the time zone on the machine running pytest and the timezone configured in the fake Home Assistant instance configured when running tests. The former is not predictable, the latter is `-07:00` unless the test is configured otherwise.

If you don't tell freezegun what timezone to use, it will assume the time zone used by the machine running pytest, which is not what you want.

The next problem is that the `schedule` enity sets up a listener for a future time, which won't fire if you travel backwards in time.

All in all, something like this should work:
```diff
diff --git a/tests/components/schedule/test_recorder.py b/tests/components/schedule/test_recorder.py
index e0ae47e19dd..6e752aad2ad 100644
--- a/tests/components/schedule/test_recorder.py
+++ b/tests/components/schedule/test_recorder.py
@@ -4,6 +4,7 @@ from __future__ import annotations

 from datetime import timedelta

+from freezegun.api import FrozenDateTimeFactory
 import pytest

 from homeassistant.components.recorder.history import get_significant_states
@@ -18,8 +19,9 @@ from tests.components.recorder.common import async_wait_recording_done


 @pytest.mark.usefixtures("recorder_mock", "enable_custom_integrations")
-async def test_exclude_attributes(hass: HomeAssistant) -> None:
+async def test_exclude_attributes(hass: HomeAssistant, freezer: FrozenDateTimeFactory) -> None:
     """Test attributes to be excluded."""
+    freezer.move_to("2024-08-02 06:30:00-07:00")  # Before Friday event
     now = dt_util.utcnow()
     assert await async_setup_component(
         hass,
@@ -50,8 +52,23 @@ async def test_exclude_attributes(hass: HomeAssistant) -> None:
     assert state.attributes[ATTR_ICON]
     assert state.attributes[ATTR_NEXT_EVENT]

+    # Move to during Friday event
+    freezer.move_to("2024-08-02 07:30:00-07:00")
+    async_fire_time_changed(hass, fire_all=True)
     await hass.async_block_till_done()
-    async_fire_time_changed(hass, dt_util.utcnow() + timedelta(minutes=5))
+    state = hass.states.get("schedule.test")
+    assert state.attributes["party_level"]
+
+    # Move to during Sunday event
+    freezer.move_to("2024-08-04 11:30:00-07:00")
+    async_fire_time_changed(hass, fire_all=True)
+    await hass.async_block_till_done()
+    state = hass.states.get("schedule.test")
+    assert state.attributes[CONF_DATA]
+
+    await hass.async_block_till_done()
+    freezer.tick(timedelta(minutes=5))
+    async_fire_time_changed(hass)
     await hass.async_block_till_done()
     await async_wait_recording_done(hass)
```

Reply:
Ahh, that explains it! I tried specifying UTC because I thought the UTC-7 was coming from my computer's time zone, so I didn't realize that was _supposed_ to be the time zone used for testing.

I just pushed that change (with some tweaks to the asserts to make them more specific). Thank you again for the help!